### ISSUE TITLE
How to block Bittorrent traffic via nDPI module (IPtables)

### ISSUE THREAD

**Issue by houmie** [XTLS/Xray-core#2300]
Most VPS service providers do not allow Bittorrent downloading due copyright issues. However, some people are not aware of this. And no matter how much this is explained there is always somebody using torrents to download movies. The best way is to stop the torrent traffic within xray-server altogether. But this is not easy.

In theory the protocol could be blocked via this config below.

Inbounds:
```
"sniffing":{
            "enabled":true,
            "destOverride":[
               "http",
               "tls"
            ]
         }
```

```
"routing":{
      "domainStrategy":"IPIfNonMatch",
      "rules":[
         {
            "type":"field",
            "outboundTag":"block",
            "protocol":[
              "bittorrent"
            ]
         }
       ]
}
```

However this doesn't work reliably with VLESS-VISION-TCP-REALITY.

Looking at the /var/log/xray/access.log I can see some packages are blocked and others go through. It can't block all of the connections, it will only slow this down.

```
2023/07/07 16:39:13 140.228.xx:32368 accepted tcp:51.159.4.xxx:55463 [block]
2023/07/07 16:39:13 140.228.xx:32012 accepted tcp:96.244.189.xxx:58946 [direct]
2023/07/07 16:39:18 140.228.xx:29768 accepted tcp:ipv6.torrent.ubuntu.com:443 [block]
2023/07/07 16:39:18 140.228.xx:29378 accepted tcp:torrent.ubuntu.com:443 [block]
```

I'm using Bittorrent app on Android (from Playstore):
![Screenshot_20230708-090420](https://github.com/XTLS/Xray-core/assets/1832933/5ca9f8d1-d6e3-4e4a-8ff6-441fe37ca1f7)

As you can see there is no way to stop the peer-to-peer traffic.

I know from experience, that this [nDPI](https://github.com/vel21ripn/nDPI) project works with Wireguard.

But because V2Ray/Xray operates at the application layer of the network stack (as opposed to the network layer for Wireguard/OpenVPN), it doesn't need to alter IP packet headers, so there's no need for NAT or masquerading. 

The core challenge here is that the nDPI module is designed to operate at the network layer, inspecting packets directly. While Xray operates at the application layer, it doesn't mean nDPI or any L7 filter couldn't analyze its traffic. Xray, as a proxy tool, encrypts the traffic between the client and the server. This makes it challenging for a DPI tool like nDPI to determine the application-level protocol, because it cannot see into the encrypted data.

Unless there was a way to hook that into the unencrypted traffic within Xray.

So what can we really do to prevent Bittorrent properly?

**Comment by houmie** [comment]
Doesn't that mean you would be using Wireguard instead of Xray though? Or are you combining an Xray protocol with Wireguard?

**Comment by timi-owo** [comment]
Blocking tracker servers would be better.

**Comment by houmie** [comment]
I have tried blocking tracker servers before, but it's not a reliable solution.  
They can disable the VPN, open the tracker site, download the magnet file and open it. Then activating VPN again and start downloading.  It's very easy.

**Comment by houmie** [comment]
@mmmray 

> I use xray, but not with the Freedom protocol, my outbound protocol is wireguard. clients still connect to xray

That's amazing. I didn't know that is even possible.  Does it mean the user can still use VLESS-VISION-TCP-REALITY to overcome the firewall and connect to your VPS. Then the outgoing protocol from your server points to your local Wireguard that diverts the traffic to an external service (like Mulvad).  That will make the connection a bit slower though, right?

And do you have any example how to set this up? This is the first time I hear about this.

**Comment by houmie** [comment]
Thank you for sharing this. That's very kind.

**Comment by donnyxray** [comment]
Alternative solution: use xray-core torrent filter to _detect_ torrent users, not to clean up traffic. Then use a simple script that tails the log file to block the user completely for a number of hours and send them an automated notice.

I've been doing it this way for years and it's very effective at avoiding DMCA notices. Most users respond very reasonably as they understand it protects a service they enjoy. Typically they just make a mistake and have no ill intent.

**Comment by houmie** [comment]
@donnyxray That's a great alternative solution. Would you mind to elaborate a bit more on how you detect if the user is using torrent via xray-core torrent filter?

```
{
        "type": "field",
        "outboundTag": "block",
        "protocol": [
          "bittorrent"
    ]
},

```

This is the one thing I know that tries to identify bittorrent to block it. But that works very poorly. Torrents can work even without a seeder and find other peers to download from via PEX and DHT.  So it is almost impossible to block it, but as you say it should be possible to detect it.

So how is your solution? Could you give a sample and elaborate a bit more in detail, please?

Thank you

**Comment by nielspeen** [comment]
I cannot copy our entire code here, but basically we just do a ```tail -f xray.log|grep block``` and then parse the client IP address and block it completely with ```iptables```.

Regular users will hit regular trackers/peer and be detected. It's very effective. 

Expert users would not use your service for torrent, unless they are purposely evil. It's unlikely. Never happened to me.

**Comment by donnyxray** [comment]
A simplified example would be:
```
tail -f /var/log/xray/access.log | grep --line-buffered '[block]' | awk '{print $3}' | cut -d':' -f1 | xargs -I {} sudo iptables -A INPUT -s {} -j DROP
```

This way it's not necessary to identify all torrent traffic. You just need to detect 1 single connection to block the user who makes a mistake. In my experience this is 100% effective in avoiding DMCA notices.

Regarding PEX, DHT, etc: I'm sure if someone nefariously wanted to shut down your VPS they could succeed. But this is an unlikely scenario. In my experience it's just people who make a mistake and are easily detected.

**Comment by kastov** [comment]
> A simplified example would be:
> 
> ```
> tail -f /var/log/xray/access.log | grep --line-buffered '[block]' | awk '{print $3}' | cut -d':' -f1 | xargs -I {} sudo iptables -A INPUT -s {} -j DROP
> ```
> 
> This way it's not necessary to identify all torrent traffic. You just need to detect 1 single connection to block the user who makes a mistake. In my experience this is 100% effective in avoiding DMCA notices.
> 
> Regarding PEX, DHT, etc: I'm sure if someone nefariously wanted to shut down your VPS they could succeed. But this is an unlikely scenario. In my experience it's just people who make a mistake and are easily detected.

Great solution, but seems like it is not work with Trojan+WS+TLS, but works with TCP.

**Comment by houmie** [comment]
Hello @donnyxray 

Sorry for replying so late. I just tried your approach to identify torrent usage, but there's an issue: nothing gets tagged as [block] because all blocked traffic is sent to the blackhole. Let me explain:

In Xray, I have the following snippet to block Bittorrent:
```
{
    "type": "field",
    "outboundTag": "block",
    "protocol": [
        "bittorrent"
    ]
},
```

Because they are tagged as [block], they will be caught by the blackhole and simply discarded. They never appear in the access.log in the first place.
```
{
    "protocol": "blackhole",
    "tag": "block",
    "settings": {}
}
```

I tested this on my Mac. I connected to my Xray VPS server and started downloading Ubuntu.iso as a torrent.

At first, nothing goes through because of the block, which is great. But if I stop the connection and let it do the handshake and start downloading a little bit, then reconnect back to the Xray VPS, the traffic will no longer be blocked. Xray can't detect this because the handshake occurred before the user reconnected.

All I see in my access.log is this:
```
2024/10/21 12:45:41 from 188.214.x.xxx:64305 accepted tcp:82.49.189.xxx:56665 [direct] email: 0b8c6154-856d-4b32-8932-8e9f10d14000@VLESS_TCP_REALITY_B.io
2024/10/21 12:45:41 from 188.214.x.xxx:65416 accepted tcp:76.155.150.xxx:11903 [direct] email: 0b8c6154-856d-4b32-8932-8e9f10d14000@VLESS_TCP_REALITY_B.io
2024/10/21 12:45:42 from 188.214.x.xxx:63070 accepted tcp:121.119.115.xxx:14664 [direct] email: 0b8c6154-856d-4b32-8932-8e9f10d14000@VLESS_TCP_REALITY_B.io
```

These are the torrent traffic entries, but all are tagged as [direct].

I understand that catching just one entry should be enough to block the traffic and prevent DMCA notices. However, I'm catching none because there are no entries with [block]. If there were, they would be filtered out in the blackhole anyway.

I'm not sure how you achieved this. Do you mind elaborating, please? I think there must be another part to your configuration that I'm missing.

**Comment by donnyxray** [comment]
Please check your own opening post. It has log lines showing **[block]**. 

I don't know what you changed to make these log lines disappear. Your original configuration will do the trick.

**Comment by UnknownDev2018** [comment]
for openvpn block torrents?

**Comment by tarmalonchik** [comment]
@houmie 
Did you manage with displaying in logs [block] tag to detect torrent traffic?
Because I noticed, this is not working with version above 1.8.24. 
I cannot understand, why routing of bittorrent traffic is not working in new versioins of xray core

**Comment by houmie** [comment]
@tarmalonchik Exactly. These logs are no longer appearing. I think it's an issue with the later Xray versions. Currently, I'm using 24.9.30. It's a shame, because it was a good workaround that Donny had suggested.

Please let me know if you have found a solution. Thank you

**Comment by freecode505** [comment]
This will avoid blocking ports oh person who plays video games only detects torrent patterns.

**Comment by donnyxray** [comment]
> I made my own deep detection system on the server, it blocks all the IP that sends p2p to my server until reaching the client IP, blocking it for 5 hours, I managed to detect all types Torrent patterns...Many people use my servers and it is effective. I was thinking of making it public, I struggled for a long time with this, you know when you make something public someone always finds vulnerability in being public ☹️

It can also result in the vulnerability being fixed very quickly.

**Comment by houmie** [comment]
@freecode505 Please make it public and share with us. Don't worry about people finding vulnerabilities. That's the beauty of open source, we help each other.

**Comment by freecode505** [comment]
Ok I will document it well but I will share it encrypted, but it really is very good, the best solution I have found, it works for any ovpn v2ray protocol etc.I will let you know

**Comment by houmie** [comment]
Thank you. I think many people would be very grateful. Don't worry, just share the solution as a public repo. If you encrypt it, then people may have problems accessing it, and it defeats the purpose of rallying the community around it.

But yes, I think documentation is very important, so everyone can understand.

On my part, I have tried everything I could with [nDPI](https://github.com/vel21ripn/nDPI), but nDPI can't penetrate the Xray protocols. I believe this is because Xray runs at the Application layer and there is no traffic being forwarded (masqueraded) unlike Wireguard or OpenVPN. This is why I gave up at the end.

Looking forward to seeing your solution.

**Comment by tarmalonchik** [comment]
@houmie I think we need to create a bug. This will be the best way to resolve this issue.

**Comment by Johnnybegood24** [comment]
> Ok I will document it well but I will share it encrypted, but it really is very good, the best solution I have found, it works for any ovpn v2ray protocol etc.I will let you know

Where could i find your solution for v2ray protocol?🙏

**Comment by UnknownDev2018** [comment]
https://github.com/UnknownDev2018/block-p2p-UDPGuard

Sorry for the delay. I've been very busy. This is my other account. I'm the same @freecode505 

Here's my free script that has managed to block 99% of torrents on my VPN users.

ENJOY!!...


@houmie @kastov @timi-owo @Fangliding @donnyxray

**Comment by houmie** [comment]
Thanks so much for sharing this @UnknownDev2018 The script looks really comprehensive and I'm very impressed by your work.

I have a question please; I believe if the Bittorrent traffic goes over https or if the clients have encryption enabled, how can this string matching work? https://github.com/UnknownDev2018/block-p2p-UDPGuard/blob/977fbe139613dbdd0d33da5833174a130f33a00d/block-p2p.sh#L159

You know what I mean? But I'm no expert, I could be wrong.

Thank you

**Comment by UnknownDev2018** [comment]
Well, I have tried with Utorrents that according to encrypts the traffic and it is still detected. I have used several clients, blocking torrents little by little, it blocks each p2p IP that shares packages or processes packages on the server, once all are blocked the download will be 0 bytes and finally blocking the IP of the user who downloads these packages. This script has several patterns to detect torrent traffic. In the same way, if you find a client that downloads without being detected, let me know and I will solve it.
